lib/srv/term.go

```diff
@@ -168,7 +178,13 @@ func (t *terminal) AddParty(delta int) {
 }
 
 // Run will run the terminal.
-func (t *terminal) Run(_ context.Context) error {
+func (t *terminal) Run(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+
 	var err error
 	// Create the command that will actually execute.
 	t.cmd, err = ConfigureCommand(t.ctx)
```

Comment:
I was also confused by it, but `t.ctx` has type `ServerContext` where `ctx` is `context.Context`. I can rename `t.ctx` to indicate that this is "not the context that you're thinking of".

Reply:
We're not using and we shouldn't. This code is executed in the main `teleport` process. If we kill/terminate the child process, we will end up in the same place that we're in right now. Currently, the tree looks like that `teleport -> teleport exec -> bash`. Our current implementation kills `teleport exec`, which causes `bash` to be re-assigned to PID 1. Calling cancel on that process would make no difference as, from what I remember, SIGTERM does nothing to `teleport exec`.
With my changes, when `teleport` want to stop the SSH session, it will inform `teleport exec` to kill `bash`. In this scenario:
1. `bash` and all subprocesses will be stopped (unless started with `nohup`)
2. `teleport exec` has a chance to run some post-process actions https://github.com/gravitational/teleport/blob/master/lib/srv/reexec.go#L411-L416 which are currently skipped

In summary, we're using `(*Process).Kill()` or (*Process).Signal()` to interact with any processes give us more flexibility and we won't end up in that weird scenario again when someone cancels context and kills the wrong process.